feat: voice rewrite guide (#595)
discordjs/guide: guide/voice/life-cycles.md

# Life cycles

Two of the main components that you'll interact with when using `@discordjs/voice` are:

> - **VoiceConnection** – maintains a network connection to the Discord voice servers

Reviewer: ```suggestion
- **VoiceConnection** – maintains a network connection to the Discord voice server(s)
```

- **AudioPlayer** – plays audio resources across a voice connection

> Both voice connections and audio players are _stateful_, and you are able to subscribe to changes in their state as they progress through their respective lifecycles.

Reviewer: ```suggestion
Both voice connections and audio players are _stateful_, and you are able to subscribe to changes in their state as they progress through their respective life cycles.
```

It's important to listen for state change events, as they will likely require you to take some action. For example, a voice connection entering the `Disconnected` state will probably require you to attempt to reconnect it.

Their individual life cycles with descriptions of their states are documented on their respective pages.

Listening to changes in the life cycles of voice connections and audio players can be done in two ways:

## Subscribing to individual events

```ts
import { VoiceConnectionStatus, AudioPlayerStatus } from '@discordjs/voice';

connection.on(VoiceConnectionStatus.Ready, (oldState, newState) => {
	console.log('Connection is in the Ready state!');
});

player.on(AudioPlayerStatus.Playing, (oldState, newState) => {
	console.log('Audio player is in the Playing state!');
});
```

:::tip

> One advantage of listening for transitions to individual states is that it becomes a lot easier to write sequential logic. This is made easy using our [state transition helper](https://github.com/discordjs/voice/blob/main/examples/basic/util.ts), found in our examples.

Reviewer 2: The link 404's.

## Subscribing to individual events

```js
const { VoiceConnectionStatus, AudioPlayerStatus } = require('@discordjs/voice');

connection.on(VoiceConnectionStatus.Ready, (oldState, newState) => {
	console.log('Connection is in the Ready state!');
});

player.on(AudioPlayerStatus.Playing, (oldState, newState) => {
	console.log('Audio player is in the Playing state!');
});
```

:::tip
One advantage of listening for transitions to individual states is that it becomes a lot easier to write sequential logic. This is made easy using our [state transition helper](https://github.com/discordjs/voice/blob/main/src/util/entersState.ts). An example is shown below.

```js
const { AudioPlayerStatus, entersState } = require('@discordjs/voice');

async function start() {
	player.play(resource);
	try {
		await entersState(player, AudioPlayerStatus.Playing, 5_000);
		// The player has entered the Playing state within 5 seconds
		console.log('Playback has started!');
	} catch (error) {
		// The player has not entered the Playing state and either:
		// 1) The 'error' event has been emitted and should be handled
		// 2) 5 seconds have passed
		console.error(error);
	}
}

> void start();

Reviewer 3: Does not seem needed in this context, and most of the guide doesn't use it so this is for consistency, perhaps...
```suggestion
start();
```

---

> forceTheme: purple

Reviewer 4: ```suggestion
pageTheme: purple
```

---

# Life cycles

Two of the main components that you'll interact with when using `@discordjs/voice` are:

- **VoiceConnection** – maintains a network connection to a Discord voice server
- **AudioPlayer** – plays audio resources across a voice connection

> Both voice connections and audio players are _stateful_, and you are able to subscribe to changes in their state as they progress through their respective life cycles.

Reviewer: ```suggestion
Both voice connections and audio players are _stateful_, and you can subscribe to changes in their state as they progress through their respective life cycles.
```
